### PR TITLE
[_transactions2] Part 30: MatrixIndexOutOfBounds (Live Reloading CellLoader Limits)

### PR DESCRIPTION
**Pull request by jeremyk-91**
**This PR has dependencies on #3857, so please review that first!**

**Goals (and why)**:
- Allow runtime configuration of the numbers the cell loader takes into account: cross-column batch limits, and load batch limits. In benchmarking 200 looked like a reasonable number, but I can't guarantee this is right for all stacks.
- Have an escape hatch for disabling cross-column batching, if necessary

**Implementation Description (bullets)**:
- Implement `CassandraCellLoadingConfig` which allows specifying what these numbers are
- Enforce some simple invariants on the config
- Wire the config through to the `CellLoader`.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added a deserialisation test and tests for config invariants. Also updated the existing CellLoadingBatcher test to accommodate batches.

**Concerns (what feedback would you like?)**:
- Should I have rolled this change into the `CassandraKeyValueServiceReloadableConfig`

**Where should we start reviewing?**: CassandraCellLoadingConfig.java

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java [hunk 1]
```diff
@@ -123,11 +140,9 @@ void loadWithTs(String kvsMethodName,
             final CassandraKeyValueServices.ThreadSafeResultVisitor visitor,
             final ConsistencyLevel consistency) {
         final ColumnParent colFam = new ColumnParent(CassandraKeyValueServiceImpl.internalTableName(tableRef));
-        CellLoadingBatcher batcher = CellLoadingBatcher.create(
-                (numRows) -> logRebatchingWarnMessage(host, tableRef, numRows));
-
         List<Callable<Void>> tasks = Lists.newArrayList();
-        for (final List<Cell> partition : batcher.partitionIntoBatches(cells)) {
+        for (final List<Cell> partition : batcher.partitionIntoBatches(
+                cells, (numRows) -> logRebatchingWarnMessage(host, tableRef, numRows))) {
```
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java]: You can pass in the Consumer into the CLB constructor, and then have partitionIntoBatches take the host and tableRef as arguments

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcher.java [hunk 2]
```diff
@@ -20,91 +20,71 @@
...
         return ImmutableList.of(cells);
     }
 
-    private static boolean shouldExplicitlyAllocateBatchToColumn(int multigetMultisliceBatchLimit, List<Cell> cells) {
-        return cells.size() > multigetMultisliceBatchLimit;
+    private static boolean shouldExplicitlyAllocateBatchToColumn(CassandraCellLoadingConfig config, List<Cell> cells) {
+        return cells.size() > config.crossColumnLoadBatchLimit();
     }
 
     private static ListMultimap<byte[], Cell> indexCellsByColumnName(Collection<Cell> cells) {
```
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcher.java]: We can lose the comment below

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcherTest.java [hunk 3]
```diff
@@ -33,37 +33,42 @@
...
     private static final int SINGLE_QUERY_LIMIT = 10;
+    private static final CassandraCellLoadingConfig LOADING_CONFIG = ImmutableCassandraCellLoadingConfig.builder()
+            .crossColumnLoadBatchLimit(CROSS_COLUMN_LIMIT)
+            .singleQueryLoadBatchLimit(SINGLE_QUERY_LIMIT)
+            .build();
 
     private IntConsumer rebatchingCallback = mock(IntConsumer.class);
-    private CellLoadingBatcher batcher = new CellLoadingBatcher(
-            () -> CROSS_COLUMN_LIMIT, () -> SINGLE_QUERY_LIMIT, rebatchingCallback);
+    private CellLoadingBatcher batcher = new CellLoadingBatcher(() -> LOADING_CONFIG);
```
Comment (gmaretic) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcherTest.java]: I'd like to see a test where we live reload the config

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraConstants.java [hunk 4]
```diff
@@ -60,6 +60,9 @@
     public static final int DEFAULT_MUTATION_BATCH_COUNT = 5000;
     public static final int DEFAULT_UNRESPONSIVE_HOST_BACKOFF_TIME_SECONDS = 30;
 
+    public static final int DEFAULT_CROSS_COLUMN_LOAD_BATCH_LIMIT = 200;
+    public static final int DEFAULT_SINGLE_QUERY_LOAD_BATCH_LIMIT = 50_000;
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraConstants.java]: Reasoning for the numbers:
- `200` was a good number based on internal benchmarking on r4.2xlarge nodes, and
- `50000` was the old behaviour of CellLoader1.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraCellLoadingConfig.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        return CassandraConstants.DEFAULT_SINGLE_QUERY_LOAD_BATCH_LIMIT;
+    }
+
+    @Value.Check
+    public void check() {
+        Preconditions.checkState(
+                crossColumnLoadBatchLimit() > 0,
+                "crossColumnLoadBatchLimit should be positive, but found %s",
+                crossColumnLoadBatchLimit());
+        Preconditions.checkState(
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraCellLoadingConfig.java]: Sure, yeah this is for the message.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcher.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcher.java]: We can - though I think it should be included, as it looks like something people might instinctively change to the simpler method without realising the implications

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java]: Sure, that's fine - have changed that. I was intending for this to be reused elsewhere (conceptually `InetSocketAddress` and `TableReference` aren't exactly things which a batcher should know about) but this should be fine for now.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraConstants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraConstants.java]: Maybe we should do some benchmarking for the latter number when we have the time

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcherTest.java [hunk 6]
```diff
@@ -113,6 +114,40 @@ public void simultaneouslyHandlesLargeColumnsAndSmallColumns() {
...
+        verify(rebatchingCallback, never()).consume(any(), any(), anyInt());
+    }
+
+    @Test
+    public void respondsToChangesInSingleQueryLimitBatchParameter() {
+        AtomicReference<CassandraCellLoadingConfig> config = new AtomicReference<>(LOADING_CONFIG);
+        CellLoadingBatcher reloadingBatcher = new CellLoadingBatcher(config::get, rebatchingCallback);
+
+        reloadingBatcher.partitionIntoBatches(rowRange(SINGLE_QUERY_LIMIT, 0), ADDRESS, TABLE_REFERENCE);
+        config.set(ImmutableCassandraCellLoadingConfig.builder()
```
Comment (gmaretic) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoadingBatcherTest.java]: Maybe do this before the method call? And extract a method

**Pull request merged**